### DISCUSSION TITLE
In an Android PWA, how do I open links in the default browser?

### DISCUSSION THREAD

**Discussion by fvsch** [elk-zone/elk#2411] [Q&A]
I installed Elk (elk.zone) as a PWA on Android via Chrome.

When I click on an external link in a toot or in the Elk UI, it opens as an overlay view powered by Chrome and sharing the same session data as the Elk PWA and/or standalone Chrome (not sure which, or if they’re one and the same).

I want to open links in the default browser instead, which in my case is Firefox for Android, but could be any browser (including standalone Chrome).

Is that doable currently? Or should I make this an issue or feature request?

**Comment by jgaehring** [comment]
Yea, I would like this, too, though I recognize others may have the opposite preference. But for me it's especially undesirable b/c of #1187.

I'm surprised how difficult it is to find info on this behavior for Android/Chrome. My first wrong assumption was this something in the `manifest.json`, something related to [`display`](https://developer.mozilla.org/en-US/docs/Web/Manifest/display) or [`display_override`](https://developer.mozilla.org/en-US/docs/Web/Manifest/display_override) perhaps. Nope. Then it seemed like the behavior could be avoided by conditionally adding/removing the `target="_blank"` attribute to `<a>` tags (#906 is not strictly related but demonstrates the basic concept), something I was hopeful could be configured in preferences and toggled by the user. As far as I know, that would still prevent a WebView from opening, at least in a native Android app, but I realized ([somehwat circuitously](https://web.dev/add-manifest/#scope)), that this might actually be an [Android Custom Tab](https://developer.chrome.com/docs/android/custom-tabs/), not a WebView. It looks like _maybe_ there was a time when you could [disable Custom Tabs from the Android System and/or Google Search settings](https://android.stackexchange.com/a/145772/26135), but as far as I can see in Android 13 (Chrome 117, Google 14), that option does not exist now, if it ever did. This seems to be a [known issue](https://stackoverflow.com/questions/46487112/chrome-61-avoid-pwa-opening-blank-links-in-custom-tab) and there has been an [open bug report on Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=1171894) since Jan 2021, which has bounced back and forth between "Untriaged" and "WontFix" status but doesn't seem to be getting any traction. That's where I finally hit a dead end.

I happily welcome others to check any and all of my assumptions here, and hopefully provide a solution b/c it totally escapes me. I just wish I hadn't wasted so much time trying unsuccessfully to figure it out. :sweat_smile:

**Comment by userquin** [comment]
Here some links, maybe we can add a preference to allow navigate in browser adding `target="_blank"` based on that preference and PWA installed (and on mobile?):
- https://developer.chrome.com/docs/web-platform/launch-handler/
- https://github.com/WICG/pwa-url-handler/blob/main/handle_links/explainer.md

Here the configuration https://github.com/elk-zone/elk/blob/main/modules/pwa/i18n.ts#L120-L123

**Comment by Zepfanman** [comment]
Don't know if I should open a new thread for this, but I'm using Mull (Firefox/Tor fork) on Android and it doesn't give an option to install as a PWA at all. I opened a similar thread here: https://gitlab.com/divested-mobile/mull-fenix/-/issues/92

**Reply by userquin** [reply]
There are browsers not supporting PWA, you should check Mull documentation. Elk requires local storage and indexed db features (and VAPID protocol for push notifications)

**Reply by userquin** [reply]
https://docs.elk.zone/pwa#install-elk-button-not-working

**Reply by Zepfanman** [reply]
Thanks. I opened the issue here, too, but not sure if there's a good resolution. https://gitlab.com/divested-mobile/mull-fenix/-/issues/92

**Comment by userquin** [comment]
That behavior is given by the web manifest, every browser can have disparate behavior opening links, I'll check if we can add or remove some flags to change the behavior.

The behavior cannot be changed since the web manifest is created at build-time.

**Comment by rvyhvn** [comment]
Not really related but I see that the share_target is already implemented but I can't find Elk entry on my share dialog when I want to share an image from gallery, browser or other apps.

Device: Android 15, HyperOS 2.0
Browser: Vivaldi

**Reply by userquin** [reply]
You must install elk pwa, once installed elk should be included in the shared app list:
https://github.com/elk-zone/elk/pull/1100

**Reply by rvyhvn** [reply]
I came here after that PR..

I already installed it but there are none and I don't know why.

**Reply by userquin** [reply]
Maybe not listed in the default lis and you need to click somewhere (more button or some silimar button). 

Note: you must be logged with an account. 

Can you check if installing elk pwa with chrome works? I have no idea about vivaldi browser.

**Reply by rvyhvn** [reply]
> the default lis and you need to click somewhere 

Pardon me, but what buttons are you referring to?

Yes, I already logged in on my browser (Vivaldi) will check on Chrome, but I don't prefer chrome as my browser if that's also your concern. If it's the browser's problem will it be supported? If no then it's unfortunate maybe Elk is not for me. Or maybe it's more to the phone's problem? HyperOS' problem perhaps?

**Reply by rvyhvn** [reply]
Yes, it shows with Chrome, so it's a browser problem.. is it possible to work with other chromium based browser other than Chrome?

**Reply by rvyhvn** [reply]
Nevermind, it's Google's fault.

https://www.reddit.com/r/vivaldibrowser/comments/1jksqhk/android_pwas_dont_get_added_to_the_app_drawer/

> Chrome can generate WebAPKs, Vivaldi and other browser cannot. Google promised to make it work for other browser, but that was years ago and it hasn't happened, yet.
